tools/tungsten/core/src/com/google/android/filament/tungsten/texture/TextureCache.kt

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (C) 2018 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.filament.tungsten.texture
+
+import com.google.android.filament.Texture
+import com.google.android.filament.tungsten.Filament
+import java.io.File
+import java.util.concurrent.CompletableFuture
+
+typealias FutureTexture = CompletableFuture<Texture>
+
+object TextureCache {
+
+    // Maps from canonical file path to Filament Texture
+    private val cache = mutableMapOf<String, FutureTexture>()
+    private var defaultTexture: Texture? = null
+
+    // Only accessed from Filament thread
+    private var allowNewTextures = true
+    private val textures = mutableListOf<Texture>()
+
+    /**
+     * Loads the image into a Filament texture and returns a future that will be completed when the
+     * texture loads successfully or completed exceptionally if an error occurs.
+     */
+    fun getTextureForFile(file: File): FutureTexture {
+        cache.computeIfAbsent(file.canonicalPath) {
+            createTextureForImageSource(file)
+        }
+        return cache[file.canonicalPath] ?: getDefaultTexture()
+    }
+
+    fun getDefaultTexture(): FutureTexture {
+        defaultTexture?.let { return CompletableFuture.completedFuture(it) }
```

Comment:
Looks like coroutines would be great for this kind of work. They should be final in Kotlin 1.3